marshaling_test.go

```diff
@@ -0,0 +1,158 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package worksheets
+
+import (
+	"encoding/json"
+
+	"github.com/stretchr/testify/require"
+)
+
+func (s *Zuite) TestMarshaling_simple() {
+	ws := defs.MustNewWorksheet("all_types")
+	forciblySetId(ws, "the-id")
+	ws.MustSet("text", NewText(`some text with " and stuff`))
+	ws.MustSet("bool", NewBool(true))
+	ws.MustSet("num_0", MustNewValue("123"))
+	ws.MustSet("num_2", MustNewValue("123.45"))
+	ws.MustSet("undefined", &Undefined{})
```

Comment:
I'm confused about why this doesn't show up in the json.

Reply:
That's because we never write undefined values in `ws.data`, absence of a value for a field indicated undefinedness.

Reply:
ah ok, that makes sense. Having a case for `*Undefined` in `marshalValue` made me think it'd be in the json too, but it sounds like that'll never happen?

Reply:
It would if you add it to a slice, because there it needs to appear... Which brings up that it should really be converted to `null` rather than `"undefined"`. Will fix now!